datascience/maps.py

```diff
@@ -0,0 +1,510 @@
+"""
+
+Integrates the folium package to draw maps.
+
+"""
+
+from IPython.core.display import HTML
+import json
+import folium
+import random
+
+#########################
+# Exposed Functionality #
+#########################
+
+
+def draw_map(center=None, **kwargs):
+    """ 
+    Draw a map with center & zoom containing all points and
+    regions that displays points as circles and regions as polygons.
+    """
+
+    return Map(
+        location=center,
+        **kwargs).map().to_html()
+
+
+#########
+# Utils #
+#########
+
+
+def _map_to_defaults(defaults, params, excludes=[], includes=[]):
+    """
+    Squishes together two dictionaries, with one containing defaults
+    """
+    rval = defaults.copy()
+    rval.update(params)
+    return {
+        k: v for k, v in rval.items() 
+        if k not in excludes
+        and k in (includes or rval.keys())
+    }
+
+
+def _hook(f):
+    """
+    Decorator for obj method, checks for PREFIX_before and PREFIX_after 
+    a method is called 
+    """
+
+    def call_hook(self, template, *args, **kwargs):
+        fn = getattr(self, template % f.__name__, None)
+        if callable(fn):
+            fn(*args, **kwargs)
+
+    def method(self, *args, **kwargs):
+        call_hook(self, '_before_%s', *args, **kwargs)
+        rval = f(self, *args, **kwargs)
+        call_hook(self, '_after_%s', *args, **kwargs)
+        return rval
+    
+    method.__name__ = getattr(f, '__name__', method.__name__)
+    return method
+
+
+class ProgrammerError(Exception):
+    """ Thrown when programmers make a mistake """
+    
+    message = ''
+    prefix = 'Please report this to your instructors'
+    
+    def get_message(self):
+        return '%s: %s' % (self.prefix, self.message)
+
+
+class MapError(Exception):
+    """ Thrown when users make a mistake """
+    pass
+
+
+######################
+# Folium Abstraction #
+######################
+
+
+class MapEntity:
+    """
+    General Map entity
+    - attributes get and set, using map[prop]
+    - all other data get and set, using map.prop
```

Comment:
I think this could be more clear as well. When I read `General Map entity` I thought this was a abstract class that only different types of `Map` inherit from. But, it looks like all our other map classes inherit from this one. So this is more like a class that helps with data getting / setting? I'm not entirely sure. I'd love to know why this class is here, as I'm sure there's a good reason but it's not immediately obvious from the docstring.

What's the difference between an attribute and `other data`?